src/mlpack/methods/ann/augmented/tasks/add_impl.hpp

```diff
@@ -0,0 +1,155 @@
+/**
+ * @file add_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of AddTask class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_AUGMENTED_TASKS_ADD_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_TASKS_ADD_IMPL_HPP
+
+#include "add.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+namespace tasks /* Task utilities for augmented */ {
+
+AddTask::AddTask(const size_t bitLen) : bitLen(bitLen)
+{
+  if (bitLen <= 0)
+  {
+    std::ostringstream oss;
+    oss << "AddTask::AddTask(): binary length (" << bitLen << ") "
+        << "is not positive!"
+        << std::endl;
+    throw std::invalid_argument(oss.str());
+  }
+}
+
+const void AddTask::Generate(arma::field<arma::mat>& input,
+                             arma::field<arma::mat>& labels,
+                             const size_t batchSize,
+                             bool fixedLength)
+{
+  arma::field<arma::vec> vecInput = arma::field<arma::colvec>(batchSize);
+  arma::field<arma::vec> vecLabels = arma::field<arma::colvec>(batchSize);
+  size_t sizeA = bitLen, sizeB = bitLen;
+  for (size_t i = 0; i < batchSize; ++i)
+  {
+    if (!fixedLength)
+    {
+      arma::vec weights(bitLen - 1);
+      // We have two binary numbers with exactly two digits (10 and 11).
+      // Increasing length by 1 double the number of valid numbers.
+      weights = arma::exp2(arma::linspace(1, bitLen - 1, bitLen - 1));
+
+      mlpack::distribution::DiscreteDistribution d(1);
+      d.Probabilities(0) = std::move(weights);
```

Comment:
You could compress this a bit, I think:

```
d.Probabilities(0) = arma::exp2(arma::linspace(1, bitLen - 1, bitLen - 1));
```

then you don't need the `weights` object. :)